### PR TITLE
snap/quota, systemd, wrappers: misc set of fixes for quota groups on Centos, AL2

### PR DESCRIPTION
**Pull request by anonymouse64**
These changes all taken together enable quota group memory usage tracking to now work correctly in conjunction with https://github.com/snapcore/snapd/pull/10316 and https://github.com/snapcore/snapd/pull/10287.

Specifically, we can avoid the bugs that version of systemd has around memory reporting by checking if the group has any tasks in it and also expanding the cases we consider a unit as inactive in IsActive. 

This should go in _before_ #10316 in order to make the spread test there pass on CentOS and AL2.

### CODE REVIEW

**Review by mvo5 — APPROVED** [review]
Thank you

File: systemd/systemd.go [hunk 1]
```diff
@@ -266,6 +266,10 @@ type Systemd interface {
 	// CurrentMemoryUsage returns the current memory usage for the specified
 	// unit.
 	CurrentMemoryUsage(unit string) (quantity.Size, error)
+	// CurrentTasksCount returns the number of tasks (processes, threads, kernel
+	// threads if enabled, etc) part of the unit, which can be a service or a
+	// slice.
+	CurrentTasksCount(unit string) (int, error)
```
Comment (bboozzoo) [inline systemd/systemd.go:272]: Negative task count does not make much sense IMO
```suggestion
	CurrentTasksCount(unit string) (uint, error)
```

File: systemd/systemd.go [hunk 2]
```diff
@@ -593,33 +597,82 @@ func (s *systemd) getGlobalUserStatus(unitNames ...string) ([]*UnitStatus, error
...
+	splitVal := strings.SplitN(cleanVal, "=", 2)
 	if len(splitVal) != 2 {
-		return 0, fmt.Errorf("invalid property format from systemd for MemoryCurrent")
+		return "", fmt.Errorf("invalid property format from systemd for %s (got %s)", key, cleanVal)
+	}
+
+	return strings.TrimSpace(splitVal[1]), nil
+}
+
+func (s *systemd) CurrentTasksCount(unit string) (int, error) {
```
Comment (bboozzoo) [inline systemd/systemd.go:618]:
```suggestion
func (s *systemd) CurrentTasksCount(unit string) (uint, error) {
```

File: systemd/systemd.go [hunk 3]
```diff
@@ -593,33 +597,82 @@ func (s *systemd) getGlobalUserStatus(unitNames ...string) ([]*UnitStatus, error
...
 	}
 
-	trimmedVal := strings.TrimSpace(splitVal[1])
+	// if the unit is inactive or doesn't exist, the tasks current can be
+	// reported as "[not set]"
+	if tasksStr == "[not set]" {
+		return 0, fmt.Errorf("tasks count unavailable")
+	}
+
+	intVal, err := strconv.Atoi(tasksStr)
```
Comment (bboozzoo) [inline systemd/systemd.go:630]: should probably use ParseUint here then

File: wrappers/services.go [hunk 4]
```diff
@@ -88,6 +88,10 @@ Before=slices.target
 # Always enable memory accounting otherwise the MemoryMax setting does nothing.
 MemoryAccounting=true
 MemoryMax=%[2]d
+
+# Always enable task accounting in order to be able to count the processes/
+# threads, etc for a slice
+TasksAccounting=true
```
[the remark this answers is not in the document]
Comment (pedronis) [reply wrappers/services.go:96]: do we know if this actually works, do we have a spread test?

File: systemd/systemd.go [hunk 5]
```diff
@@ -593,33 +597,82 @@ func (s *systemd) getGlobalUserStatus(unitNames ...string) ([]*UnitStatus, error
...
+	if err != nil {
+		return 0, fmt.Errorf("invalid property value from systemd for TasksCurrent: cannot parse %q as an integer", tasksStr)
+	}
+
+	return intVal, nil
+}
+
+func (s *systemd) CurrentMemoryUsage(unit string) (quantity.Size, error) {
+	memStr, err := s.getPropertyStringValue(unit, "MemoryCurrent")
```
Comment (bboozzoo) [inline systemd/systemd.go:639]: I played a bit with this on CentOS 7, so far the findings suggest that when MemoryLimit is not set (meaning there is no group under memory controller hierarchy), MemoryCurrent reported by systemd has an absurd value in exabytes. At least adding MemoryAccounting=true makes the problem go away, as the group is created and memory stats are there. So the bogus value is observed only when there's no kernel source for the statisics and seems to be some internal default value in systemd. I was able to reproduce the same exact behavior for TasksCurrent, that is setting TasksAccounting=true makes the reported value be sane, while restarting the slice without accounting, is again a very high default value.

I suspect that the systemd behavior may have changed and with unified hierarchy a process is always part of some slice which has memory & task accounting enabled (which afaik is always true with cgroup v2).

Bit erratic log from my experiments: https://paste.ubuntu.com/p/TfYRHQTqWn/

Comment (pedronis) [reply systemd/systemd.go:639]: it sound like we should set MemoryAccounting=true in our slices anyway?

**Comment by mvo5** [comment]
Thanks for the PR and the comments - if supporting Centos7 and AL2 is a big support issue at this point, let's simply error here and not support systemd versions that are this old. If we have real demand for this feature we can always go back and see if we can work with these old versions of systemd (or lobby for a newer systemd in there images :) - let avoid sinking too much time into these old systemds.

File: snap/quota/quota.go [hunk 6]
```diff
@@ -103,6 +103,21 @@ func (grp *Group) CurrentMemoryUsage() (quantity.Size, error) {
 		return 0, nil
 	}
 
+	// also check how many tasks are in the group, it could exist and be active,
+	// but not have any active services running in it, and with systemd 219 on
+	// Amazon Linux 2 / Centos 7, there is a bug where the current memory usage
+	// for a unit in such a state is reported as being 16 exbibytes which is too
+	// large to fit into a 32-bit int and is obviously wrong (it should be 0)
+	numTasks, err := sysd.CurrentTasksCount(grp.SliceFileName())
```
Comment (bboozzoo) [inline snap/quota/quota.go:111]: Hm with MemoryAccounting=true|yes I only see a bogus value for a slice that is inactive, but we exit on this earlier. Once the slice is active, even if there are no tasks, the reported value seems to be correct.

File: systemd/systemd.go (hunk 5 above)
Comment (anonymouse64) [reply systemd/systemd.go:639]: but we have always set MemoryAccounting=true in the slice units we generate ?

**Comment by anonymouse64** [comment]
the ordering of these pr's has gotten too muddled so I will close these and reopen them in a more logical splitting/ordering
but to be clear, there is now a functional spread test on master we can use/modify to demonstrate whether changes are working or not

**Pull request closed without merging**